src/Effects/Reverse.cpp

```diff
@@ -0,0 +1,218 @@
+    memory.dealloc(reverterR);
+}
+
+//Cleanup the effect
+void Reverse::cleanup(void)
+{
+    reverterR->reset();
+    reverterL->reset();
+}
+
+
+//Effect output
+void Reverse::out(const Stereo<float *> &input)
+{
+    // prepare the brocessing buffers
+    if(Pstereo) //Stereo
+        for(int i = 0; i < buffersize; ++i) {
+            efxoutl[i] = input.l[i] * pangainL;
+            efxoutr[i] = input.r[i] * pangainR;
+        }
+    else //Mono
+        for(int i = 0; i < buffersize; ++i)
+            efxoutl[i] = (input.l[i] * pangainL + input.r[i] * pangainR);
+
+    // process external timecode to sync
+
+    unsigned int beat_new = 0;
+    if (time->tempo && speedfactor && PsyncMode == HOST)
+    {
+        tick = (time->beat-1)*1920 + time->tick;
+        const unsigned int delay_ticks = int(1920.0f * speedfactor);
+        beat_new = tick/delay_ticks;
+        const unsigned int phase_ticks = tick%delay_ticks;
+
+        if(beat_new!=beat_new_hist) {
+            const float syncPos = (phase_ticks/delay_ticks)*(60.0f/time->tempo)*(float)(time->samplerate());
+            reverterL->sync(syncPos);
+            if(Pstereo) reverterR->sync(syncPos);
+        }
+    }
+    // store beat_new for next cycle
+    beat_new_hist = beat_new;
+
+    // do the actual processing
+    reverterL->filterout(efxoutl);
+    if(Pstereo) reverterR->filterout(efxoutr);
+    else memcpy(efxoutr, efxoutl, bufferbytes);
+}
+
+void Reverse::update()
+{
+    // process noteon trigger
+    if( (PsyncMode == NOTEON || PsyncMode == NOTEONOFF) ) {
+        reverterL->sync(0.0f);
+        if(Pstereo) reverterR->sync(0.0f);
+    }
+}
+//Parameter control
+void Reverse::setvolume(unsigned char _Pvolume)
+{
+    Pvolume = _Pvolume;
+
+    if(insertion == 0) {
+        if (Pvolume == 0) {
+            outvolume = 0.0f;
+        } else {
+            outvolume = powf(0.01f, (1.0f - Pvolume / 127.0f)) * 4.0f;
+        }
+        volume    = 1.0f;
+    }
+    else
+        volume = outvolume = Pvolume / 127.0f;
+    if(Pvolume == 0)
+        cleanup();
+}
+
+void Reverse::setdelay(unsigned char value)
+{
+    Pdelay = limit(value,static_cast<unsigned char>(0),static_cast<unsigned char>(127));
+    reverterL->setdelay(float(Pdelay+1)/128.0f*MAX_REV_DELAY_SECONDS);
+    reverterR->setdelay(float(Pdelay+1)/128.0f*MAX_REV_DELAY_SECONDS);
+}
+
+void Reverse::setphase(unsigned char value)
+{
+    Pphase = value;
+    reverterL->setphase(float(Pphase)/127.0f);
+    reverterR->setphase(float(Pphase)/127.0f);
+}
+
+void Reverse::setcrossfade(unsigned char value)
+{
+    Pcrossfade = value;
+    reverterL->setcrossfade(float(value)/100.0f);
+    reverterR->setcrossfade(float(value)/100.0f);
+}
+
+void Reverse::setsyncMode(unsigned char value)
+{
+    PsyncMode = value;
+    reverterL->setsyncMode((SyncMode)value);
+    reverterR->setsyncMode((SyncMode)value);
+}
+
+unsigned char Reverse::getpresetpar(unsigned char npreset, unsigned int npar)
+{
+    return 0;
```

Comment:
Looks like presets aren't implemented. Even two would make this work about the same as other effects.

Reply:
Or, if you don't want to add any presets, I would recommend using unnamed parameters, or using `(void)npreset` etc - to silence a compiler warning.

Reply:
i added presets now.